router/failed-events-manager.go

```diff
@@ -0,0 +1,177 @@
+package router
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+)
+
+var (
+	failedEventsManager FailedEventsManagerI
+)
+
+type FailedEventRowT struct {
+	DestinationID string
+	RecordID      json.RawMessage
+}
+
+var (
+	failedKeysTablePrefix  = "failed_keys"
+	failedKeysExpire       = 48 * time.Hour
+	failedKeysCleanUpSleep = 24 * time.Hour
+	failedKeysEnabled      = false
+)
+
+type FailedEventsManagerI interface {
+	SaveFailedRecordIDs(map[string][]*FailedEventRowT, *sql.Tx)
+	DropFailedRecordIDs(jobRunID string)
+	FetchFailedRecordIDs(jobRunID string) []*FailedEventRowT
+}
+
+type FailedEventsManagerT struct {
+	dbHandle *sql.DB
+}
+
+
+
+func GetFailedEventsManager() FailedEventsManagerI {
+	if failedEventsManager == nil {
+		fem := new(FailedEventsManagerT)
+		dbHandle, err := sql.Open("postgres", jobsdb.GetConnectionString())
+		if err != nil {
+			panic(err)
+		}
+		fem.dbHandle = dbHandle
+		failedEventsManager = fem
+	}
+
+	return failedEventsManager
+}
+
+func (fem *FailedEventsManagerT) SaveFailedRecordIDs(taskRunIDFailedEventsMap map[string][]*FailedEventRowT, txn *sql.Tx) {
+	if !failedKeysEnabled {
+		return
+	}
+	for taskRunID, failedEvents := range taskRunIDFailedEventsMap {
+		table := fmt.Sprintf(`%s_%s`, failedKeysTablePrefix, taskRunID)
+		sqlStatement := fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s (
+		destination_id TEXT NOT NULL,
+		record_id JSONB NOT NULL,
+		created_at TIMESTAMP NOT NULL);`, table)
+		_, err := txn.Exec(sqlStatement)
+		if err != nil {
+			panic(err)
+		}
+		insertQuery := fmt.Sprintf(`INSERT INTO %s VALUES($1, $2, $3);`, table)
+		stmt, err := txn.Prepare(insertQuery)
+		if err != nil {
+			panic(err)
+		}
+		createdAt := time.Now()
+		for _, failedEvent := range failedEvents {
+			if len(failedEvent.RecordID) == 0 || !json.Valid(failedEvent.RecordID) {
+				pkgLogger.Infof("skipped adding invalid recordId: %s, to failed keys table: %s", failedEvent.RecordID, table)
+				continue
+			}
+			_, err = stmt.Exec(failedEvent.DestinationID, failedEvent.RecordID, createdAt)
+			if err != nil {
+				panic(err)
+			}
+		}
```

Comment:
minor: Would a batch update query be efficient? If so can we track as backlog task?

Reply:
Added a task
https://www.notion.so/rudderstacks/cfe8e8ef60a34485a373e789b7b6933a?v=a07a0b1350f84d46af9792260b39a194&p=278a8bbf4c1546aa942d6cf171dd168a